make/libraries

```diff
@@ -65,6 +65,28 @@ LIBIDAS = $(IDAS)/lib/libsundials_nvecserial.a $(IDAS)/lib/libsundials_idas.a
 STAN_IDAS_HEADERS := $(shell find stan -name *idas*.hpp)
 $(STAN_IDAS_HEADERS) : $(LIBIDAS)
 
+##
+# Build rules for MPI
+##
+
+$(BOOST)/stage/lib/mpi.so:
+	@mkdir -p $(dir $@)
+	cd $(BOOST); ./bootstrap.sh
+	cd $(BOOST); echo "using mpi ;" >> user-config.jam
+	cd $(BOOST); ./b2 --user-config=user-config.jam --with-mpi --with-serialization
+	@if [ "$(OS_TYPE)" = "mac" ]; then\
```

Comment:
So what follows here (the install_name_tool commands) are commands that need to be run on a Mac when we're building `mpi.so`, but not necessarily when we're compiling code against `mpi.so`?

If this is correct I think this structure makes sense. But if it were the case that the `install_name_tool` commands mentioned were needed any time you were on a Mac and had `STAN_MPI` defined, I'd be inclined to put them in the `os_mac` file.

Reply:
It took me while starring at the documentation of macOS linker tools and stuff on the net to figure this out. I never intended to dive this deep into it, but now it works. Should we collect more dynamic libraries in stan-math which we want to distribute, we may revisit as to how we set the rpath logic.